crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+            return Err(Errno::Notdir);
+        }
+
+        let host_path = path::from_host(po_path.as_os_str())?;
+        let host_path_len = host_path.len().try_into()?;
+
+        if host_path_len > path_len {
+            return Err(Errno::Nametoolong);
+        }
+
+        trace!("     | (path_ptr,path_len)='{}'", host_path);
+
+        let mut bc = GuestBorrows::new();
+        let path_arr = path.as_array(host_path_len);
+        let slice = path_arr.as_raw(&mut bc)?;
+        unsafe { &mut *slice }.copy_from_slice(host_path.as_bytes());
+
+        Ok(())
+    }
+
+    fn fd_pwrite(
+        &self,
+        fd: types::Fd,
+        ciovs: &types::CiovecArray<'_>,
+        offset: types::Filesize,
+    ) -> Result<types::Size> {
+        trace!(
+            "fd_pwrite(fd={:?}, ciovs={:?}, offset={})",
+            fd,
+            ciovs,
+            offset,
+        );
+
+        let mut slices = Vec::new();
+        let mut total_len: types::Size = 0;
+        let mut bc = GuestBorrows::new();
+        bc.borrow_slice(ciovs)?;
+        for ciov_ptr in ciovs.iter() {
+            let ciov_ptr = ciov_ptr?;
+            let ciov: types::Ciovec = ciov_ptr.read()?;
+            let buf = ciov.buf;
+            let buf_len = ciov.buf_len;
+            total_len = total_len.checked_add(buf_len).ok_or(Errno::Inval)?;
+            let as_arr = buf.as_array(buf_len);
+            let as_raw = as_arr.as_raw(&mut bc)?;
+            slices.push(unsafe { &*as_raw });
+        }
+
+        let entry = unsafe { self.get_entry(fd)? };
+        let file = entry
+            .as_descriptor(
+                types::Rights::FD_WRITE | types::Rights::FD_SEEK,
+                types::Rights::empty(),
+            )?
+            .as_file()?;
+
+        if offset > i64::max_value() as u64 {
+            return Err(Errno::Io);
+        }
+
+        let mut buf = Vec::with_capacity(total_len as usize);
+        for slice in &slices {
+            buf.extend_from_slice(slice);
+        }
+        let host_nwritten = match file {
+            Descriptor::OsHandle(fd) => fd::pwrite(&fd, &buf, offset)?,
+            Descriptor::VirtualFile(virt) => virt.pwrite(buf.as_mut(), offset)?,
+            _ => {
+                unreachable!(
+                    "implementation error: fd should have been checked to not be a stream already"
+                );
+            }
+        };
+        trace!("     | *nwritten={:?}", host_nwritten);
+        let host_nwritten = host_nwritten.try_into()?;
+
+        Ok(host_nwritten)
+    }
+
+    fn fd_read(&self, fd: types::Fd, iovs: &types::IovecArray<'_>) -> Result<types::Size> {
+        trace!("fd_read(fd={:?}, iovs={:?})", fd, iovs);
+
+        let mut bc = GuestBorrows::new();
+        let mut slices: Vec<&'_ mut [u8]> = Vec::new();
+        bc.borrow_slice(&iovs)?;
+        for iov_ptr in iovs.iter() {
+            let iov_ptr = iov_ptr?;
+            let iov: types::Iovec = iov_ptr.read()?;
```

Comment:
This `read` is unsafe in combination with the long-lived borrows here, you'll need to add the `GuestPtr<Iovec>` to the `GuestBorrows` to make sure this pointer doesn't overlap with previous buffers.

Reply:
Hmm, out of curiosity, is the example [here](https://github.com/bytecodealliance/wasmtime/blob/master/crates/wiggle/tests/wasi.rs#L130) wrong then?

Reply:
Hmm, but wouldn't `borrow_slice` take care of it for us here?

Reply:
Oh no nevermind my interpretation is wrong, I think this should be fine.